### PR TITLE
[core] move function and actor importer away from pubsub

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

This PR moves function import to a lazy way. Several benefits of this:

- worker start up is faster since it doesn't need to go through all functions exported
- gcs pressure is smaller since 1) we don't need to export key and 2) all loads are done when needed.
- get rid of function table channel

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by iycheng** [comment]
@ericl could you help check the PR and let me know whether I missed something? Hopefully, it should work.

If this is working, I'll get rid of run_on_all_nodes and then import_thread.py is not necessary anymore.

File: python/ray/worker.py [hunk 1]
```diff
@@ -1635,7 +1635,7 @@ def connect(
         # are the same.
         # When using an interactive shell, there is no script directory.
         if not interactive_mode:
-            script_directory = os.path.abspath(os.path.dirname(sys.argv[0]))
```
Comment (iycheng) [inline python/ray/worker.py:1638]: This is a bug and somehow not triggered before this PR.
sys.argv[0] is always `python`

File: python/ray/tests/test_basic_5.py [hunk 2]
```diff
@@ -124,34 +124,37 @@ def test_internal_kv(ray_start_regular):
...
+from filelock import FileLock
 from pathlib import Path
+import pickle
+
+lock_file = "{str(lock_file)}"
+data_file = Path("{str(data_file)}")
 
 def init_func(worker_info):
-    a = ray.get_actor("recorder", namespace="n")
-    a.record.remote(worker_info['worker'].worker_id)
```
Comment (iycheng) [inline python/ray/tests/test_basic_5.py:147]: The old code has a race condition. Before the worker calls run in loop, we can't call any remote function. 

If init_func is called before the worker calls run in loop, it'll crash.
If init_func is called after the worker executes tasks, it'll be wrong.

**Comment by iycheng** [comment]
This PR also makes me aware it's not that easy to support create new worker when GCS is down.

File: python/ray/_private/function_manager.py [hunk 3]
```diff
@@ -389,14 +389,22 @@ def _wait_for_function(self, function_descriptor, job_id, timeout=10):
...
+                        break
+                    else:
+                        key = make_function_table_key(
+                            b"RemoteFunction",
+                            job_id,
+                            function_descriptor.function_id.binary(),
+                        )
+                        if self.fetch_and_register_remote_function(key) is True:
+                            break
+                elif not self._worker.actor_id.is_nil():
```
Comment (mwtian) [inline python/ray/_private/function_manager.py:403]: Just `else`?

File: python/ray/_private/function_manager.py [hunk 4]
```diff
@@ -389,14 +389,22 @@ def _wait_for_function(self, function_descriptor, job_id, timeout=10):
...
+                if self._worker.actor_id.is_nil():
+                    if function_descriptor.function_id in self._function_execution_info:
+                        break
+                    else:
+                        key = make_function_table_key(
+                            b"RemoteFunction",
+                            job_id,
+                            function_descriptor.function_id.binary(),
+                        )
+                        if self.fetch_and_register_remote_function(key) is True:
```
Comment (mwtian) [inline python/ray/_private/function_manager.py:401]: Does this busy loop if fetching fails?

File: python/ray/_private/import_thread.py [hunk 5]
```diff
@@ -43,6 +43,8 @@ def __init__(self, worker, mode, threads_stopped):
         self.num_imported = 0
         # Protect writes to self.num_imported.
         self._lock = threading.Lock()
+        # Do importing if initialized first
```
Comment (mwtian) [inline python/ray/_private/import_thread.py:46]: First among ...?

File: python/ray/tune/examples/pbt_dcgan_mnist/pbt_dcgan_mnist_func.py [hunk 6]
```diff
@@ -95,7 +95,9 @@ def dcgan_train(config, checkpoint_dir=None):
         "--data-dir", type=str, default="~/data/", help="Set the path of the dataset."
     )
     args, _ = parser.parse_known_args()
-    ray.init()
+
+    path = os.path.dirname(os.path.abspath(__file__))
```
Comment (mwtian) [inline python/ray/tune/examples/pbt_dcgan_mnist/pbt_dcgan_mnist_func.py:99]: Why is this change needed?

**Review by mwtian — COMMENTED** [review]
Have we ran nightly tests on this? Moving importing to lazy may have unknown risks.

**Comment by mwtian** [comment]
Great, really hope this can work!

**Comment by iycheng** [comment]
@mwtian it's running here https://buildkite.com/ray-project/release-tests-pr/builds/1232
So far it looks good.

**Comment by ericl** [comment]
+1 to trigger nightly tests on this. I'm fine with merging this if nightly tests pass 3x in a row with no import related issues or hangs, and also making sure the unit tests are not flaky.

**Comment by iycheng** [comment]
Win tests are broken but not related to the flow change. Most are because the CI is not written well. I'll fix that and run nightly test again.

File: python/ray/_private/function_manager.py (hunk 4 above)
Comment (iycheng) [reply python/ray/_private/function_manager.py:401]: It's ok since the worker will not make progress anyway before

**Comment by iycheng** [comment]
https://buildkite.com/ray-project/release-tests-pr/builds/1370

**Comment by iycheng** [comment]
https://buildkite.com/ray-project/release-tests-pr/builds/1387

**Comment by iycheng** [comment]
Merge to master and trigger the third run here.

https://buildkite.com/ray-project/release-tests-pr/builds/1232
https://buildkite.com/ray-project/release-tests-pr/builds/1370
https://buildkite.com/ray-project/release-tests-pr/builds/1387

I feel there might not be enough resources. Let's wait and see.

**Comment by iycheng** [comment]
scheduling_test_many_0s_tasks_many_nodes also failed with master
https://buildkite.com/ray-project/release-tests-branch/builds/423#5c52eb26-7f90-4db1-a15e-a92dbb1eed11

shuffle_1tb_5000_partitions_1650980795 also failed with master
https://console.anyscale.com/o/anyscale-internal/projects/prj_2xR6uT6t7jJuu1aCwWMsle/clusters/ses_fZiUMTm2bMUMVHFSdvJJSqg3

**Review by mwtian — APPROVED** [review]
Looks good. One comment is that we are still importing all dependencies in their export order, by calling `_do_import()` on worker startup and when a dependency is not found. This PR is not changing to per-function lazy import yet, so there is no import ordering issue. Maybe it is useful to clarify in the PR description.

**Comment by mwtian** [comment]
Chatted with Yi offline, the only purpose of `_do_import()` now is running functions that need to run on all workers. We are indeed importing directly from GCS KV with `fetch_and_register_remote_function()`.

**Comment by iycheng** [comment]
@ericl  it seems like lacking of resources and queues forever. The first one has completely finished. The second/third one is partially finished. Do you think it's good to let it go?

**Comment by ericl** [comment]
We can give it a try and keep an eye out.
